### PR TITLE
Introduce behat tests

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Introduce new domain tests structure, setup, scripts
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10733
| How to test?  | Travis tests should be green and include the new unit test

⚠️ This needs to add a new Travis build.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12090)

### CODE REVIEW

File: tests/Integration/Behaviour/features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
+<?php
+
+use Behat\Behat\Context\Context;
+use LegacyTests\PrestaShopBundle\Utils\DatabaseCreator;
+
+/**
+ * Defines application features from the specific context.
+ */
+class FeatureContext implements Context
```
Comment (eternoendless) [inline tests/Integration/Behaviour/features/bootstrap/FeatureContext.php]: Maybe this should be a base class (without traits) and then subclass according to the testing context?

File: tests/Integration/Behaviour/features/sql_manager.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
+Feature: SQL Manager
```
Comment (eternoendless) [inline tests/Integration/Behaviour/features/sql_manager.feature]: I wonder if `feature` file names should be in snake case or camel case.

**Comment by eternoendless** [comment]
Seems OK to me for a PoC, the structure will need to evolve as we add more tests

File: tests/Integration/Behaviour/features/sql_manager.feature (hunk 2 above)
Comment (matks) [reply tests/Integration/Behaviour/features/sql_manager.feature]: It's up to use to decide. I'm used to snake case names for non-php files (YAML config files, XML config files, bash scripts ...) but it makes no difference for usage.

**Comment by Quetzacoalt91** [comment]
> :warning: This needs to add a new Travis build.

The only thing you have to do is updating the file `.travis.yml`. Depending on your needs, you won't modify the same part of the file.

* If you want to run your integration tests on all defined PHP versions (at this time: 5.6 + 7.2), add the job here:

```yml
env:
  global:
    - SYMFONY_DEPRECATIONS_HELPER=disabled
  matrix:
    - PRESTASHOP_TEST_TYPE=lint
    - PRESTASHOP_TEST_TYPE=unit
    - PRESTASHOP_TEST_TYPE=integration # <-- added here
    - PRESTASHOP_TEST_TYPE=e2e
```

* If you want to add only one job, with a specific version of PHP, add it in the `include` property:

```yml
matrix:
  include:
    - php: 7.2 # Or another PHP version you'd prefer
      env: PRESTASHOP_TEST_TYPE=integration
   [...] # Existing rules
```

**Comment by matks** [comment]
@tomlev @eternoendless 
Requested changs have been applied along a refactor.

Here's what I did:
- rename folders to Features and Context in order to allow Composer PSR-4 autoloading in order to use namespace for FeatureContexts
- split the bootstrap into a dedicated `bootstrap.php` file
- improved the README
- created an `AbstractPrestaShopFeatureContext` that has the boilerplate code necessary to run behat tests for prestashop (application setup, database reset capabilities, cache clear capabilities ...)
- all FeatureContext should extend it, and this starts with the `SqlManagerFeatureContext` which is not a trait anymore
- whenever we add new Contests (1 feature / domain = 1 Context) such as CustomerFeatureContext, CartFeatureContext to store new steps they must be added to the [behat.yml](https://github.com/PrestaShop/PrestaShop/pull/12090/files#diff-59e5e15b1fc334d1b63da8ef7f3f1338) file

File: tests/Integration/Behaviour/features/bootstrap/FeatureContext.php (hunk 1 above)
Comment (matks) [reply tests/Integration/Behaviour/features/bootstrap/FeatureContext.php]: Done, see https://github.com/PrestaShop/PrestaShop/pull/12090#issuecomment-453058815

File: tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php [hunk 3]
```diff
@@ -48,12 +54,12 @@ public function assertStoredSqlRequestCount($count)
 
         $realCount = current($realCountResults)['result'];
 
-        if ((int) $realCount !== (int) $count) {
+        if ((int)$realCount !== (int)$count) {
```
Comment (eternoendless) [inline tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php]: This change will fail the linter check, the white space must be reinstated.

File: tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * @Given there is :count stored SQL requests
+     * @Then there should be :arg1 stored SQL request
+     */
+    public function assertStoredSqlRequestCount($count)
+    {
+        // @todo: need improvement and decoupling
+        $legacyDatabaseSingleton = \Db::getInstance(_PS_USE_SQL_SLAVE_);
+        $realCountResults = $legacyDatabaseSingleton->executeS('SELECT COUNT(*) AS result FROM ps_request_sql');
```
Comment (PierreRambaud) [inline tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php:53]: use DB_PREFIX instead?

File: tests/Integration/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+### PrestaShop behat tests
+
+Feature files are stored in `Features` folder.
+
+FeatureContext files are stored in `Features/Context` folder. They must extend
+the `AbstractPrestaShopFeatureContext` that provide the setup necessary to perform
+tests on PrestaShop.
+
+[1]: http://docs.behat.org/en/v2.5/guides/1.gherkin.html
```
Comment (PierreRambaud) [inline tests/Integration/README.md:42]: Can you name links instead. Easier to follow and understand.

File: tests/check_integration.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+composer run-script integration-domain-tests --timeout=0;
+DOMAIN=$?
+
+if [[ "$DOMAIN" == "0" ]]; then
+  echo -e "\e[92mINTEGRATION TESTS OK"
```
Comment (PierreRambaud) [inline tests/check_integration.sh:7]: Forget to clear color.

File: tests/check_integration.sh [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+else
+  echo -e "\e[91mINTEGRATION TESTS FAILED"
```
Comment (PierreRambaud) [inline tests/check_integration.sh:9]: Forget to clear color.

**Comment by PierreRambaud** [comment]
@matks Nice work! But can you use namespaces everywhere?

File: composer.json [hunk 8]
```diff
@@ -186,6 +187,9 @@
             "@composer create-test-db",
             "@php -d date.timezone=UTC ./vendor/bin/phpunit -c tests-legacy/old-controllers.xml"
         ],
+        "integration-behaviour-tests": [
+            "@php -d date.timezone=UTC ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml"
```
Comment (mickaelandrieu) [inline composer.json:191]: isn't Behavior instead? I'm in doubt here ^^

File: tests/Integration/Behaviour/Features/Context/AbstractPrestaShopFeatureContext.php [hunk 9]
```diff
@@ -0,0 +1,120 @@
+<?php
+
```
Comment (mickaelandrieu) [inline tests/Integration/Behaviour/Features/Context/AbstractPrestaShopFeatureContext.php:2]: File header

File: tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php [hunk 10]
```diff
@@ -0,0 +1,118 @@
+<?php
+
```
Comment (mickaelandrieu) [inline tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php:2]: file header

File: tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+/**
+ * SqlManagerFeatureContext provides behat steps to perform actions related to prestashop SQL management
+ * and validate returned outputs
+ */
+class SqlManagerFeatureContext extends AbstractPrestaShopFeatureContext
+{
+    /**
+     * @When I request the database fields from table :tableName
+     */
+    public function getDatabaseTableFieldsList($tableName)
```
Comment (mickaelandrieu) [inline tests/Integration/Behaviour/Features/Context/SqlManagerFeatureContext.php:20]: I'm not sure the function is specific to SQL Manager: shouldn't be located in the Abstract class?

**Review by mickaelandrieu — COMMENTED** [review]
Good job @matks!

File: composer.json (hunk 8 above)
Comment (PierreRambaud) [reply composer.json:191]: Behavior is the preferred spelling in American English. Behaviour is preferred everywhere else. Both are valid.

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
There are comments :)

**Comment by PierreRambaud** [comment]
Because #12634 is base on your PR and we start to add comments, we should continue this discussion in the new PR and close this one if you are ok @matks

**Comment by matks** [comment]
I'm OK with that 😄

**Pull request closed without merging**